### ISSUE TITLE
Allow api to take pre-bundled metadata information

### ISSUE THREAD

**Issue by xlc** [polkadot-js/api#829]
Currently every new instance of Api will query metadata information and then it will become usable.
This adds a significantly initialization time which is not ideal. Image someone wants to develop a serverless lambda with polkadot.js.
It will be good if the constructor of Api takes a metadata json object and if it presents, use it instead of fetch from server.
May optionally still fetch from server and print a warning if mismatch found.

**Comment by jacogr** [comment]
As stand-along instances, I believe this goes against the grain - we are already not quite updating metadata when it changes on-chain, which is an issue. (e.g. https://github.com/polkadot-js/api/issues/449) 

If you are running multiple instances inside a single container, the `.clone()` allows you to pass through metadata.

**Comment by xlc** [comment]
Outdated metadata should not cause issue. Of cause this is up to runtime developer to ensure backward compatibility. But for trivial upgrades, like add new module add new method/storage/event, it should preserve backward compatibility.

This also make it possible to create & sign transaction offline.

**Comment by KarishmaBothara** [comment]
Plan to resolve the above issue as. 
In the API constructor (packages/api/src/Base.ts) if we can pass hex string of runtimeMetaData and its related runtimeVersion as part of ApiOptions and pass it via init method to loadMeta function, where we check if the runtimeVersion passed is same as **this._rpcBase.chain.getRuntimeVersion()** if yes then we assign this.runtimeMetaData to the one passed by user or else get it from the chain i.e


```
private async loadMeta (runtimeVersion, rpcdata ): Promise<boolean> {
   if (!this._options.source || !this._options.source._isReady) {
      this._runtimeVersion = await this._rpcBase.chain.getRuntimeVersion();
      const metadata = new Metadata(rpcdata).asV0;
      this._runtimeMetadata =  this._runtimeVersion === runtimeVersion ?  metadata : await this._rpcBase.state.getMetadata();
   }
}

```
Request your feedback

**Comment by xlc** [comment]
It may help a bit but does not fundamentally solve my issue.

The current API requires a message roundtrip before it is useable. It can take few seconds on mobile network. Replace it with another simple API only reduce the time a bit but won't help with the long latency in mobile app.

And it doesn't help for the offline usage.

**Comment by jacogr** [comment]
Don't misread my comments above as "this can't work or shouldn't get it", rather read it as "this is a very dangerous and bad idea if you don't know what you are doing".

So what does the API does on init, so you only seem to cover one part -

- load meta, used for decoding
- retrieves runtime version, used in some areas for feature detection (and should be used to re-load meta when it is updated as well - not in as of yet)
- retrieves genesis hash, used for signing (immortal)

Basically, with the clone additions, you can already do what you are asking for without any additional code - on the constructor options, pass a `source` object that contains all items 3 above. However if you aim to use this with static inits, I would suggest that we get #449 in ASAP so detection can be done.

**Comment by ianhe8x** [comment]
I reckon this as a performance improvement, and `source` is not really fix because it's introduced to provide feature of clone. Currently runtimeVersion is not included in Metadata so we can not simply put metadata in source and leave version empty, if we want to achieve what Karishma proposed. 
I think what we want to achieve is assuming upgrade of runtime is not very frequently, in most cases client don't have to request metadata from remote, and even if the shipped metadata is not latest, we can detect via runtimeVersion and query it as before. This change will be totally transparent to user and reduce startup time.

**Comment by jacogr** [comment]
Closing, done in #850
